pkg/controller/namespace/deletion/namespaced_resources_deleter.go

```diff
@@ -563,3 +567,69 @@ func (d *namespacedResourcesDeleter) estimateGracefulTerminationForPods(ns strin
 	}
 	return estimate, nil
 }
+
+// updateStatusCondition does best effort to update status condition for
+// a namespace if namespace deletion didn't finish. In case of a failure, it
+// does not propagate its error further, instead it logs it and the error
+// from namespace deletion is propagated.
+func (d *namespacedResourcesDeleter) updateStatusCondition(nsName string, deletionErr error) {
+	err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		ns, err := d.nsClient.Get(nsName, metav1.GetOptions{})
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		if hasChanged := changeStatusConditionWithErr(ns, deletionErr); hasChanged {
+			_, err = d.nsClient.UpdateStatus(ns)
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't update status condition for namespace %q: %v", nsName, err))
+	}
+}
+
+// changeStatusConditionWithErr reflects the error on namespace status condition
+func changeStatusConditionWithErr(ns *v1.Namespace, err error) bool {
+	if estimate, ok := err.(*ResourcesRemainingError); ok {
+		return changeStatusCondition(ns, v1.NamespaceDeletionEstimate, "RemainingResources",
+			fmt.Sprintf("There are remaining resources, pending deletion estimate %vs", estimate.Estimate))
+	} else {
+		return changeStatusCondition(ns, v1.NamespaceDeletionFailure, "DeletionFailed",
+			fmt.Sprintf("There are remaining resources, deletion error: %v", err))
```

Comment:
I recommend not having this function and figuring out the conditions you want before even calling `updateStatusCondition`. For example, if 3 things can't be deleted, I think it'd be good if the error message called out the number of things that can't be deleted. It should also probably mention e.g. the alphabetically first one.

I think the code as written will have a lot of churn, randomly switching the message if multiple things can't be deleted?

Reply:
after doing a little bit more testing, that is exactly what is happening. Alphabetical order sounds like a good solution, thanks!